### PR TITLE
Pipeline for text model

### PR DESCRIPTION
**Pull request by janmalyy**
I developed a testing pipeline for textual models to explore various configurations of embeddings, architectures, and their parameters.

I've assigned all of you to the code review—do it whoever you want or nobody:)

I plan to include the evaluation of results later. It should only require a few lines of code, so I created the pull request now to give you some time to review the code before the lesson.

### CODE REVIEW

File: porcupaine/textual_model/PyTorchBinaryClassifier.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+import torch
```
Comment (janrygl) [inline porcupaine/textual_model/PyTorchBinaryClassifier.py]: file naming: lowecase
PyTorch is only tech
nn_binary_classifier.py?

File: porcupaine/textual_model/PyTorchBinaryClassifier.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            batch_size (int): Batch size for training.
+            epochs (int): Number of training epochs.
+        """
+        super().__init__()
+        self.lr = lr
+        self.batch_size = batch_size
+        self.epochs = epochs
+        self.classes_ = np.array([0, 1])      # attribute needed for GridSearchCV; TODO replace with not-hard-coded solution
+
+    @abstractmethod         # TODO Is this declaration properly used? I hope so but I am not sure.
```
Comment (janrygl) [inline porcupaine/textual_model/PyTorchBinaryClassifier.py]: it is ok

File: porcupaine/textual_model/PyTorchBinaryClassifier.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+                outputs = self.forward(batch_features).squeeze()
+
+                batch_class_weights = class_weights[batch_labels.long()]
+                loss = weighted_bce_loss(outputs, batch_labels.float(), weight=batch_class_weights)
+
+                loss.backward()
+                optimizer.step()
+
+                epoch_loss += loss.item()
```
Comment (janrygl) [inline porcupaine/textual_model/PyTorchBinaryClassifier.py]: what is a purpose of a variable which gets overwritten and it is not used?

File: porcupaine/textual_model/PyTorchBinaryClassifier.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    def predict(self, X):
+        """
+        Predict binary class labels.
+        Args:
+            X (pd.DataFrame): Input features.
+        Returns:
+            np.ndarray: Predicted labels.
```
Comment (janrygl) [inline porcupaine/textual_model/PyTorchBinaryClassifier.py]: describe labels here, give

Examples: section?

File: porcupaine/textual_model/PyTorchBinaryClassifier.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        Args:
+            X (pd.DataFrame): Input features.
+        Returns:
+            np.ndarray: Probabilities for the positive class.
+        """
+        self.eval()
+        with torch.no_grad():
+            X_tensor = torch.tensor(X.values, dtype=torch.float32)
+            outputs = self.forward(X_tensor)
+        return outputs.numpy()
```
Comment (janrygl) [inline porcupaine/textual_model/PyTorchBinaryClassifier.py]: give example again

File: porcupaine/textual_model/evaluate_results.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+from porcupaine.settings import *
+
+if __name__ == "__main__":
+    """
+    Creates a nice diagram where all metrics from the best models are plotted.
+    """
+
+    data_dict = {}
+    for file_name in os.listdir(RESULTS_DIR):
```
Comment (janrygl) [inline porcupaine/textual_model/evaluate_results.py]: RESULTS_DIR should have func RESULTS_DIR.iterdir()

File: porcupaine/textual_model/evaluate_results.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            file_path = os.path.join(RESULTS_DIR, file_name)
+
+            with open(file_path, 'r') as json_file:
+                json_data = json.load(json_file)
+                data_dict[file_name] = json_data["results"]
+
+    only_the_best = {}
+    for model_name, key in data_dict.items():
+        if key["f1_score_all"] > 0.65 and key["accuracy_all"] > 0.65\
+                and key["precision_0"] != 0 and key["precision_1"] != 0 and key["recall_0"] != 0 and key["recall_1"] != 0:
```
Comment (janrygl) [inline porcupaine/textual_model/evaluate_results.py]: `a != 0` is the same as `a`

Comment (janrygl) [reply porcupaine/textual_model/evaluate_results.py]: you can wrap in () for better readability

File: porcupaine/textual_model/evaluate_results.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+            only_the_best[model_name] = key
+
+    print(len(only_the_best))
```
Comment (janrygl) [inline porcupaine/textual_model/evaluate_results.py]: missing semantic context in print

File: porcupaine/textual_model/evaluate_results.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    precision_0 = [only_the_best[key]["precision_0"] for key in only_the_best]
```
Comment (janrygl) [inline porcupaine/textual_model/evaluate_results.py:29]: if you store variables into dict, you can save a lot of code

File: porcupaine/textual_model/text_model.py [hunk 10]
```diff
@@ -1,3 +1,5 @@
+# This Python file uses the following encoding: utf-8
```
Comment (janrygl) [inline porcupaine/textual_model/text_model.py]: all new Python files should use utf-8

**Review by janrygl — APPROVED** [review]
Was checking only readability, not functionality. Looks nice.
